### PR TITLE
Fix torch.fft.fftfreq writing wrong values and clobbering memory with a non-contiguous out=

### PR DESCRIPTION
**Pull request by vineethsaivs**
`fft_fftfreq_out` slices the negative-frequency half with an explicit `end` of `0`:

```cpp
auto right_slice = out.slice(0, (n + 1) / 2, 0);
```

`Tensor::slice` clamps `end` up to `start`, so this is an empty view, not a to-the-end
sentinel. `at::arange_out(right_slice, -(n/2), 0, 1)` then finds a 0-element output where it
needs `n/2` elements, resizes it, and a resize of an empty view produces a fresh contiguous
run over the storage starting at that view's offset.

For a contiguous `out` the fresh run happens to land exactly on the right elements, which is
why this has gone unnoticed. For a non-contiguous `out` it lands on the wrong ones, and it
runs off into storage that is not part of `out` at all.

## Repro

```python
import torch

tbl = torch.zeros(8, 3)
torch.fft.fftfreq(8, out=tbl[:, 0])

print(tbl[:, 0].tolist())
# [0.0, 0.125, 0.25, 0.375, -0.5, -0.125, 0.75, 0.875]     wrong
print(torch.fft.fftfreq(8).tolist())
# [0.0, 0.125, 0.25, 0.375, -0.5, -0.375, -0.25, -0.125]   expected

print(tbl[:, 1].tolist())
# [0.0, 0.0, 0.0, 0.0, -3.0, 0.0, 0.0, 0.0]                clobbered, was all zeros
```

Filling one column of a preallocated table is ordinary usage, and a non-contiguous `out` is a
supported case that `test_ops.py::test_out` exercises generally. The second effect is the worse
one: `tbl[:, 1]` is a tensor the caller never passed to the op. The value left there is unscaled
because the trailing `out.mul_(1.0 / (n * d))` walks only `out`'s own strides, so the write went
somewhere `mul_` does not reach.

## Fix

Drop the `end` argument so the slice really does run to the end of `out`. `right_slice` then has
`numel() == n/2`, which matches what `arange_out` computes, so no resize happens and the values
are written through `out`'s real strides.

The existing `test_fftfreq_out` still passes: it uses a 0-dim `out`, and `at::arange_out(out, n)`
on the line above resizes that to a contiguous `(n,)` before the slice is taken.

## Test

`test/test_spectral_ops.py::TestFFT::test_fftfreq_out_noncontiguous` builds `out` as every other
element of a `-1`-filled buffer, checks the values against the functional result, and checks that
the interleaved elements are still `-1`. It covers `fftfreq` and `rfftfreq` over both even and odd
`n`. It fails on the current `main` on both assertions and passes with this change.

I do not have a machine that can build ATen, so I validated the fix by transcribing
`fft_fftfreq_out` line for line into Python and re-registering it through the real dispatcher with
`torch.library.Library("aten", "IMPL")`. The transcription of the current code reproduces the
wrong values and the out-of-bounds write above exactly; with the one-line change it matches
`torch.fft.fftfreq` for every case in the new test. CI is the real check here, and I will follow up
on anything it turns up.

This is self-found rather than tied to an existing issue. Happy to open one first if you would
prefer that route.

cc @mruberry

### CODE REVIEW

**Comment by vineethsaivs** [comment]
This is ready for review, not a work in progress. My account currently gets a permissions error from GitHub both on opening a non-draft PR against this repo and on marking this one ready, so it opened as a draft and I cannot flip the flag myself. If a maintainer marks it ready, or is happy to review it as is, I am ready to go on CI feedback.

**Comment by vineethsaivs** [comment]
@pytorchbot label "release notes: python_frontend" "module: fft" "module: correctness (silent)"

Tagging it user facing rather than `topic: not user facing`: `torch.fft.fftfreq(out=)` is public API, and the wrong values it writes into a non-contiguous `out` are silent.
